src/systems/audio-zones-system.js

```diff
@@ -0,0 +1,161 @@
+// We apply the most restrictive audio parameters
+function paramsReducer(acc, curr) {
+  if (acc === null) acc = curr;
+  acc.gain = Math.min(acc.gain, curr.gain);
+  acc.maxDistance = Math.min(acc.maxDistance, curr.maxDistance);
+  acc.refDistance = Math.min(acc.refDistance, curr.refDistance);
+  acc.rolloffFactor = Math.max(acc.rolloffFactor, curr.rolloffFactor);
+  acc.coneInnerAngle = Math.min(acc.coneInnerAngle, curr.coneInnerAngle);
+  acc.coneOuterAngle = Math.min(acc.coneOuterAngle, curr.coneOuterAngle);
+  acc.coneOuterGain = Math.min(acc.coneOuterGain, curr.coneOuterGain);
+  return acc;
+}
+
+/**
+ * This system updates the audio-zone-sources audio-params based on the audio-zone-listener position.
+ * On every tick it computes the audio-zone-source and audio-zone-listener positions to check
+ * if the listener is inside/outside an audio-zone and applies the zone's audio parameters based.
+ * It only updates in case there has been any changes in the sources or listener's positions.
+ * If several audio-zones are in between the audio-zone-listener and the audio-zone-source, the applied
+ * audio parameters is a reduction of the audio-zones most restrictive audio parameters.
+ * i.e. If there are two audio-zones in between the listener and the source and the first one has gain == 0.1
+ * and the other has gain == 1.0, gain == 0.1 is applied to the source.
+ */
+export class AudioZonesSystem {
+  constructor(scene) {
+    this.scene = scene;
+    this.listener = null;
+    this.sources = [];
+    this.zones = [];
+    this.initialized = false;
+    this.ray = new THREE.Ray();
```

Comment:
It looks like a lot of these properties are not actually persistent properties but just used to avoid creating garbage. For those I usually tend to just either declare the variables at the root of the file, or close over them for just the function that uses them. That way any time you see `this` you know its some sort of persistent state that is going to be sticking around.

Reply:
I've moved this to a closure.